### PR TITLE
feat: add recovery codes

### LINKED ISSUE #824: 2FA Add recovery codes
Linked to #164 

We should add recovery codes, in case of the user lost his device, or cannot reach it.

### PR DESCRIPTION
**Pull request by asbiin**
This fix #824

- [x] Controller to generate codes
- [x] UI to generate codes (with vue call)
- [x] UI to copy/paster codes
- [x] Page to login with a recovery code, link from the login page

### Checklist

- [x] [CHANGELOG](https://github.com/monicahq/monica/blob/master/CHANGELOG) entry added, if necessary, under `UNRELEASED`.
- [x] If it's relevant and worth mentioning, create a changelog entry for this change. The changelog entry will appear inside the UI for all users to see. To know if your change is worth the creation of a changelog entry, [read the documentation](https://github.com/monicahq/monica/blob/master/docs/administrators/tips.md#when-is-it-relevant-to-create-a-changelog-entry).

### CODE REVIEW

File: app/Http/Controllers/Settings/RecoveryCodesController.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        // Generate new codes
+        $this->generate();
+
+        $codes = auth()->user()->recoveryCodes()->get();
+
+        return new RecoveryCodeCollection($codes);
+    }
+
+    /**
+     * Get list of recovery codes (with POST request).
```
Comment (djaiss) [inline app/Http/Controllers/Settings/RecoveryCodesController.php]: You mean?
```suggestion
     * Get list of recovery codes (with GET request).
```

File: app/Http/Controllers/Settings/RecoveryCodesController.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (count($codes) == 0) {
+            $this->generate();
+            $codes = auth()->user()->recoveryCodes()->get();
+        }
+
+        return new RecoveryCodeCollection($codes);
+    }
+
+    private function generate()
```
Comment (djaiss) [inline app/Http/Controllers/Settings/RecoveryCodesController.php]:
```suggestion
    /**
     * Generate new recovery codes.
     * @return void
     */
    private function generate()
```

File: app/Http/Resources/Settings/RecoveryCode/RecoveryCode.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+     * @param  \Illuminate\Http\Request
+     * @return array
+     */
+    public function toArray($request)
+    {
+        return [
+            'id' => $this->id,
+            'object' => 'recoverycode',
+            'recovery' => $this->recovery,
+            'used' => (bool) $this->used,
```
Comment (djaiss) [inline app/Http/Resources/Settings/RecoveryCode/RecoveryCode.php]: This is meant to be used in the API at some point right? Why don't you include the account information like the other resources then?

File: config/auth.php [hunk 4]
```diff
@@ -104,4 +104,20 @@
         ],
     ],
 
+    /*
+    |--------------------------------------------------------------------------
+    | Recovery codes generate
```
Comment (djaiss) [inline config/auth.php]:
```suggestion
    | Recovery codes generation
```

File: database/migrations/2018_11_30_154729_recovery_codes.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('recovery_codes', function (Blueprint $table) {
+            $table->increments('id');
+            $table->unsignedInteger('account_id');
+            $table->unsignedInteger('user_id');
+            $table->string('recovery', 20);
```
Comment (djaiss) [inline database/migrations/2018_11_30_154729_recovery_codes.php]: Are you sure about the `20` here? If the recovery code is configurable in the `auth` file, we might have code way longer than this. Either you put the default limit for a string, or you remove the code to be configurable (imho).

File: public/changelog.json [hunk 6]
```diff
@@ -1,5 +1,10 @@
 {
     "entries": [
+        {
+            "date": "Dec 02, 2018",
+            "title": "Security: recovery codes",
+            "description": "In the [security tab of settings](/settings/security), you can now generate recovery codes. Recovery codes are usefull if you have activated Two Factor Authentication, but you cannot use it temporarely. With a recovery code you can still login, using this one time code, to bypass Two Factor Authentication. Be carefull to save them in a secure place. ![image](/img/changelogs/2018-12-02-recovery-codes.png)"
```
Comment (djaiss) [inline public/changelog.json]:
```suggestion
            "description": "In the [security tab of settings](/settings/security), you can now generate recovery codes. Recovery codes are useful if you have activated Two Factor Authentication but can not login with it for some reasons. With a recovery code you can bypass Two Factor Authentication using this one time code. Be careful though - those codes should be stored in a secure place as they allow someone to gain access to your account. ![image](/img/changelogs/2018-12-02-recovery-codes.png)"
```

File: resources/lang/en/settings.php [hunk 7]
```diff
@@ -101,6 +101,15 @@
...
     'u2f_error_timeout' => 'Timeout reached before request could be satisfied.',
 
+    'recovery_title' => 'Recovery codes',
+    'recovery_show' => 'Get recovery codes',
+    'recovery_copy_help' => 'Copy codes in your clipboard',
+    'recovery_help_intro' => 'These are your recovery codes:',
+    'recovery_help_information' => 'You can use each recovery code once.',
+    'recovery_clipboard' => 'Codes copied in the clipboard',
+    'recovery_generate' => 'Generate new codes...',
+    'recovery_already_used_help' => 'Already used code',
```
Comment (djaiss) [inline resources/lang/en/settings.php]:
```suggestion
    'recovery_already_used_help' => 'This code has already been used',
```

File: resources/views/settings/security/u2f-enable.blade.php [hunk 8]
```diff
@@ -40,17 +40,17 @@
...
+            <a href="{{ route('security.index') }}" class="btn">{{ trans('app.cancel') }}</a>
 
           </div>
         </div>
       </div>
     </div>
   </div>
 </div>
-<script src="{{ mix('js/u2f-api.js') }}" type="text/javascript"></script>
+<script src="{{ asset(mix('js/u2f-api.js')) }}" type="text/javascript"></script>
```
Comment (djaiss) [inline resources/views/settings/security/u2f-enable.blade.php]: 👍

**Comment by djaiss** [comment]
The feature is awesome. In the Settings page, it's well done.
However I don't see the link on the login page to use one of my code.

**Comment by asbiin** [comment]
Thanks for the review!
At first I've intended to allow a user to *login* with a recovery code, but now it can be used only if 2FA is enabled, in case you can't use 2FA.
On a second though, we should be able to use a recovery code instead of password to login, but it's more sensitive attempt...

**Comment by djaiss** [comment]
Oh I see. I didn't understand this. I'll try again.

File: resources/views/settings/security/u2f-enable.blade.php (hunk 8 above)
Comment (asbiin) [reply resources/views/settings/security/u2f-enable.blade.php]: :+1:

File: database/migrations/2018_11_30_154729_recovery_codes.php (hunk 5 above)
Comment (asbiin) [reply database/migrations/2018_11_30_154729_recovery_codes.php]: OK I've removed the limit here.

File: app/Http/Resources/Settings/RecoveryCode/RecoveryCode.php (hunk 3 above)
Comment (asbiin) [reply app/Http/Resources/Settings/RecoveryCode/RecoveryCode.php]: OK the resource is not useful, I have removed it

**Comment by asbiin** [comment]
photo_2018-12-09_23-16-57

File: config/auth.php [hunk 9]
```diff
@@ -104,4 +104,20 @@
         ],
     ],
 
+    /*
+    |--------------------------------------------------------------------------
+    | Recovery codes generation
+    |--------------------------------------------------------------------------
+    |
```
Comment (djaiss) [inline config/auth.php]:
```suggestion
    | Recovery codes can be used to access your account in the event you lost
    | access to the device used to generate or receive two-factor
    | authentication codes. 
```

File: CHANGELOG [hunk 10]
```diff
@@ -14,6 +14,7 @@ UNRELEASED CHANGES:
 * Fix avatar display in searches
 * Fix conversation add/update using contact add/update flash messages
 * Fix incompatibility of people search queries with PostgreSQL
+* Add ability to generate recovery codes in order to bypass 2FA/U2F
```
[the remark this answers is not in the document]
Comment (asbiin) [reply CHANGELOG]: Alright

File: config/auth.php (hunk 9 above)
Comment (asbiin) [reply config/auth.php]: Thanks

**Pull request merged**